### ISSUE TITLE
Per-stub delays

### ISSUE THREAD

**Issue by tommysitu** [SpectoLabs/hoverfly#807]
**Is your feature request related to a problem? Please describe.**
Delays setting in `globalActions` is not flexible enough, you can only filter request by `urlPattern` or `httpMethod`. You can't filter by request body or headers for example. The main issue is that when you need granular control of delays, you have to write a lot of urlPatterns while requestMatcher already does it for you. 

**Describe the solution you'd like**
You should be able to add delay to individual request-response pair (stub) directly. 
For fixed delay:

```
"response": {
	"status": 200,
	"body": "slow response",
        "fixedDelay": 1000
}
```

For lognormal delay: 

```
"response": {
	"status": 200,
	"body": "slow response",
        "logNormalDelay": {
               "min": 100,
               "max": 10000,
               "mean": 5000,
               "median": 500
            }
}
```

Local delays (per-stub delays) should override global delays if they're happened to match on the same request. Multiple delays definitions on the same level should be additive.

**Comment by UnasZole** [comment]
I have a similar requirement, where I need my virtual service to reproduce the real response time observed on the real service for each different request.

An ideal solution would be the following : 
- Set a delay for each stub
- Allow setting a global delayMultiplier (0 = respond immediately for all requests; 1 = respond with the specified delay)

**Comment by ns3777k** [comment]
@tommysitu I think I'm gonna implement this. Only simulate and spy modes should support this? What did you mean by`Multiple delays definitions on the same level should be additive.`' ?

**Comment by tommysitu** [comment]
That would be very helpful! Feel free to have a go.

1. The delay is applied here: https://github.com/SpectoLabs/hoverfly/blob/9708a10025b41cb68e01f43dd45e9f6d3cfc0ca7/core/hoverfly.go#L209

so it supports any mode except `capture` 

2. Exactly. If the user provides both delays,  the actual delay will be the total. 
So the following is supported: 
```
"response": {
	"status": 200,
	"body": "slow response",
        "fixedDelay": 1000,
        "logNormalDelay": {
               "min": 100,
               "max": 10000,
               "mean": 5000,
               "median": 500
            }
}
```

**Comment by ns3777k** [comment]
Hmm. Taken `SynthesizeMode` for example. It doesn't grab a response from the storage but instead it calls a middleware that returns the response. So in this case, a middleware should be able to return `fixedDelay` and `logNormalDelay`?

On the hand, `SimulateMode` can grab `fixedDelay` and `logNormalDelay` from the storage in case a simulation was imported earlier and then it calls a middleware that can also return different `fixedDelay` and `logNormalDelay`. Should we apply both delay sets here?

**Comment by tommysitu** [comment]
you're right. it's possible that middleware adds the delay settings. so perhaps the `response, err := mode.Process(req, requestDetails)` should also returns a  `*ResponseDelay` struct for the specific response? if it's nil, then use global delay? otherwise, execute the per-stub delay?

**Comment by ns3777k** [comment]
Probably, I'll look into it more deeply :+1: But the question remains :-) In simulate mode during `Process` call I can have one pair of `fixedDelay` and `logNormalDelay` from the storage and the one more pair from the middleware. Should I sum them up? Or prefer one over the other?

**Comment by tommysitu** [comment]
the `RequestResponsePair` could only have one response even after applying the middleweare. so it won't end up with two `fixedDelay`.

**Comment by ns3777k** [comment]
Oh. Indeed. my bad :smile:

**Comment by tommysitu** [comment]
Thanks @ns3777 👍  and sorry for the wait. Released in [v1.3.0](https://github.com/SpectoLabs/hoverfly/releases/tag/v1.3.0)
